owner-onboarding-server/src/ov_management/mod.rs

```diff
@@ -0,0 +1,256 @@
+            }
+            Err(_) => {},
+        }; 
+    }
+
+    if !unknown_device.is_empty() {
+        return  ErrorMsg::generate(ErrorCode::UnknownDevice, unknown_device);
+    }
+
+    for i in uuids {
+        user_data
+        .ownership_voucher_store
+        .destroy_data(&Guid::from_str(&i).unwrap())
+        .await.unwrap();
+    }
+
+    Ok(Box::new(warp::reply()))
+}
+
+
+const ACCEPTED_CONTENT_TYPE:[&str; 2] = ["application/x-pem-file","application/cbor"];
+
+#[derive(Clone)]
+#[derive(Serialize,Deserialize)]
+#[derive(Eq,PartialEq,Hash)]
+#[derive(Error,Debug)]
+enum ErrorCode {
+    #[error("error")]
+    ParseError,
+    #[error("parsed")]
+    InvalidNumberOfVouchers,
+    #[error("unowned_voucher")]
+    UnwonedVoucher,
+    #[error("X-Number-Of-Vouchers")]
+    VoucherCountError,
+    #[error("required")]
+    RequireContentTypeHeader,
+    #[error("required")]
+    NoOwnershipVoucherFound,
+    #[error("unknown")]
+    UnknownDevice,
+    #[error("{0}")]
+    InvalidVoucherSignatureDescription(String),
+    #[error("failed at index")]
+    InvalidVoucherSignature,
+    #[error("{0}")]
+    IncompleteVoucherDescription(String),
+    #[error("parsed correctly")]
+    IncompleteVoucher,
+    #[error("{0}")]
+    PemParseErrorDescription(String),
+    #[error("parsed correctly")]
+    PemParseError,
+}
+
+#[derive(Serialize)]
+struct ErrorMsg {
+    error_code: Option<ErrorCode>,
+    error_details: serde_json::Value
+}
+
+trait Response<T> {
+    fn generate(e: ErrorCode,value: T ) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection>;
+    fn generate_with_description(e: ErrorCode,value: T,description: &str) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection>;
+}
+
+impl<T:Serialize> Response<T> for ErrorMsg
+{
+    fn generate(e: ErrorCode,value: T) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection> 
+    where T: Serialize {
+          let m =     ErrorMsg {
+                error_code:  Some(e.clone()),
+                error_details: json!({e.to_string(): value})
+            };
+        Ok(Box::new(warp::reply::with_status(warp::reply::json(&m), StatusCode::BAD_REQUEST)))
+    }
+    
+    fn generate_with_description(e: ErrorCode,value: T, description: &str) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection> 
+    where T: Serialize {
+          let m=     ErrorMsg {
+                error_code:  Some(e.clone()),
+                error_details: json!({
+                    "description" : description,
+                    e.to_string(): value
+                })
+            };
+        Ok(Box::new(warp::reply::with_status(warp::reply::json(&m), StatusCode::BAD_REQUEST)))
+    }
+
+}
+
+fn validate_ov<'a> (ov: &'a OwnershipVoucher,
+    owner_key: &'a PKey<Private>,
+) -> Result<bool,ErrorCode> {
+        ov.iter_entries()
+        .map_err(|e| ErrorCode::PemParseErrorDescription(e.to_string()))?
+        .into_iter().next()
+        .unwrap()
```

Comment:
I think this would panic if it's an OV without any entries?

Reply:
Its handled in line:75

Reply:
No, that is no Ownership Vouchers. This comment was about an Ownership Voucher that had no ownership transfer entries.

Reply:
Updated